i18n: load the locale depending on user preferences

dachary opened the pull request:
## Status

Ready for review

## Description of Changes

Set the locale of the source or journalist interface as follows, by
order of precedence:

  - l request argument or session['locale']
  - browser suggested locale via Accept-Languages
  - config.DEFAULT_LOCALE
  - 'en_US'

When the l request argument is specified, the locale is saved in the
session. If the l request argument is empty, the locale from the session
is cleared and the Accept-Languages header takes precedence again.

The locale requested with the l argument or via the Accept-Languages
header is matched against the available locales, as found in the
translations directory. The babel.core.negotiate_locale function is used
to find a match and it will, for instance, match:

   fr to fr_FR
   fr-FR to fr_FR
   zh-CN to zh_Hans_CN (script part)
   zh_Hans to zh_Hans

In order to improve the chances to find a match, the Accept-Languages
values are normalized via the core.Locale.parse function which knowns
about aliases, default script for Chinese etc.


## Testing

* pytest --page-layout tests/pages-layout and visually verify all pages look as they should
* pytest tests will run the additional tests that should could all modified lines
<pre>
modified   securedrop/tests/pytest.ini
@@ -1,4 +1,4 @@
 [pytest]
 testpaths = . functional
 usefixtures = setUptearDown
-addopts = --cov=../securedrop/
+addopts = --cov-report term-missing --cov=../securedrop/
</pre>
will help verify this is the case.

## Deployment

The code makes use of new config variables. It will do nothing and behave in a backward compatible way if they are not present. This should be verified by upgrading from a 0.4.3 fresh install.

## Checklist

### If you made changes to the app code:

- [x] Unit and functional tests pass on the development VM

### If you made changes to the system configuration:

- [x] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

dachary commented:
@redshiftzero this is the last one before we can merge all gettext() wrapping of strings. It is more subtle than it looks because it was squashed and patched at least a dozen times over the past two months. The rest will be easier ;-)

dachary commented:
<pre>
i18n.py                                   38      0     18      0   100%
</pre>

The claimed decreased coverage is a mystery to me...

dachary commented:
https://circleci.com/gh/freedomofpress/securedrop/2046?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
"Timeout when waiting for search string OpenSSH in ec2-52-53-233-88.us-west-1.compute.amazonaws.com:22"
</pre>
re-launching

heartsucker commented on securedrop/i18n.py:
```diff
@@ -0,0 +1,76 @@
+def setup_app(app):
+    global babel
+    babel = Babel(app)
+    assert 1 == len(list(babel.translation_directories))
```
Pedantic: Yoda assertion (`expected == test_case` and not `test_case == expected`)

dachary replied:
The **1** is indeed what is expected. Am I missing something ?

heartsucker replied:
I think traditionally it's written `some_operation_or_var == constant`.

dachary replied:
ah, ok, will change that

heartsucker commented on securedrop/i18n.py:
```diff
@@ -0,0 +1,76 @@
+            pass
+    if 'l' in request.args:
+        if len(request.args['l']) == 0:
+            if 'locale' in session:
```
I feel like we shouldn't pop this off the session until after we've gotten an acceptable alternative. Is there a reason you're popping it early? If so, can you put that in a comment.

dachary replied:
in which case do we not have an acceptable alternative ?

heartsucker replied:
None, actually, but since this function doesn't also set the `session['locale']`, I feel like it shouldn't modify it. Also, neither the J nor S app sets the session locale. They only assign it to `g`.

dachary replied:
>  since this function doesn't also set the session['locale']
but it does, in the else part of the if

heartsucker commented on securedrop/i18n.py:
```diff
@@ -0,0 +1,76 @@
+
+
+def setup_app(app):
+    global babel
```
So looking at this, it looks like we never use the `babel` object outside this function, so to avoid us adding more global state, I think it's safe to not save the output of `Babel(app)` anywhere. If we have to save it later, we should attach it to the `app` object itself so that it's unique per app instance to help with future testing.

dachary replied:
Oh, nice ! It was needed in the beginning and then never questioned it again. Removed.

heartsucker commented on securedrop/tests/test_i18n.py:
```diff
@@ -0,0 +1,148 @@
+import version
+
+
+class TestI18N(object):
```
So, so approve of everything about these tests. :D

heartsucker commented:
Checked the screenshots. Things look good.

redshiftzero commented on securedrop/tests/test_i18n.py:
```diff
@@ -0,0 +1,148 @@
+    @classmethod
+    def teardown_class(cls):
+        reload(journalist)
+        reload(source)
```
Thought: we could have a test to guard against forgetting that `DEFAULT_LOCALE` may not be set in `config.py`, something along the lines of:

```
def test_verify_default_locale_en_us_if_not_defined_in_config(self):
        DEFAULT_LOCALE = config.DEFAULT_LOCALE
        try:
            config.DEFAULT_LOCALE = None
            with app.test_client() as c:
                c.get('/')
                assert session.get('locale') == 'en_US'
                assert not_translated == gettext(not_translated)
        finally:
            config.DEFAULT_LOCALE = DEFAULT_LOCALE
```

Note: I have not tested this and it may need some modification, just jotting down the general idea

dachary replied:
Good idea ! To be perfectly honest I thought about adding a similar test and conveniently forgot about it ... Your scaffold was useful and adapted. Repushed !

redshiftzero commented on securedrop/tests/test_i18n.py:
```diff
@@ -0,0 +1,148 @@
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') is None
```
Why would we want `locale` to be `None` and not `en_US` here?

dachary replied:
Because the session is not always set. It is set when **l=XXX** and unset when **l=**

dachary commented:
@heartsucker is there anything else you need modified ?

heartsucker commented:
@dachary 

The docs say:

```python
"""
Get the locale as follows, by order of precedence:
- l request argument
- browser suggested locale, from the Accept-Languages header
- config.DEFAULT_LOCALE
- app.config['BABEL_DEFAULT_LOCALE']
"""
```

But the value is also pulled from the session in some cases. The docs should be updated to reflect that (and possibly a test to ensure a request with a session containing a locale retains that value on the next request.

dachary commented:
> But the value is also pulled from the session in some cases. The docs should be updated to reflect that 

Good catch ! I updated as follows:

    - l request argument or session['locale']
    - browser suggested locale, from the Accept-Languages header
    - config.DEFAULT_LOCALE
    - 'en_US'

with both l and session on the same line since they have equal precedence because they really are the same thing. And changed

   - app.config['BABEL_DEFAULT_LOCALE']

to 

    - 'en_US'

The former was true in a previous version of the implementation but it was decided to move to something explicit and easier to control from the user / developer perspective.

> (and possibly a test to ensure a request with a session containing a locale retains that value on the next request.

I believe the current tests verify that, in particular:

<pre>
            c.get('/?l=fr_FR', headers=Headers([('Accept-Language', 'en_US')]))
            assert session.get('locale') == 'fr_FR'
            assert translated_fr == gettext(not_translated)

            c.get('/', headers=Headers([('Accept-Language', 'en_US')]))
            assert session.get('locale') == 'fr_FR'
            assert translated_fr == gettext(not_translated)
</pre>

heartsucker commented:
> with both l and session on the same line since they have equal precedence because they really are the same thing.

But they're not, right? Someone could start a session with `de-DE` and switch to `fr-FR` via the `l` query parameter. Session should be below `l`, no?

Anyway, that's me being pedantic. I'm marking this as accepted since this change is minimal.

dachary commented:
> But they're not, right? Someone could start a session with de-DE and switch to fr-FR via the l query parameter. Session should be below l, no?

Well yes, in a way your right because the session cannot override l while the other is true. But my thinking was that l will always be the same as session['locale']. There are no situation where l and session['locale'] are different. A contrario the header can be different from the session and the session will have precedence. 

But I agree with you, it is a fine point and thanks for agreeing to something that is debatable :-)
